### PR TITLE
Feature: Pay endpoint

### PR DESCRIPTION
**Pull request by Tigh-Gherr**
Assuming there are two instances of payd and p4 running, and one is funding, and an invoice exists on the receiver, you can call:

```
curl -H 'Content-Type: application/json' http://payd:8443/api/v1/pay -d '{
        "payToURL": "http://p4-merchant:28445/api/v1/payment/2JV7g4n"
}'
```

### CODE REVIEW

File: service/pay.go [hunk 1]
```diff
@@ -0,0 +1,210 @@
...
+
+	spvEnvelope, err := p.spvc.CreateEnvelope(ctx, tx)
+	if err != nil {
+		return nil, errors.Wrapf(err, "failed to create spv envelope for tx %s", tx.String())
+	}
+
+	ack, err := p.p4.PaymentSend(ctx, req, payd.PaymentSend{
+		SPVEnvelope: spvEnvelope,
+		ProofCallbacks: map[string]payd.ProofCallback{
+			"https://" + p.svrCfg.Hostname + "/api/v1/proofs/" + spvEnvelope.TxID: {},
```
Comment (theflyingcodr) [inline service/pay.go:168]: This should prob be a new config for P4 server or something like that.

File: service/pay.go [hunk 2]
```diff
@@ -0,0 +1,210 @@
...
+	if err != nil {
+		return nil, errors.Wrapf(err, "failed to create ec private key for script %s", script.String())
+	}
+
+	return &bt.LocalSigner{
+		PrivateKey: privKey,
+	}, nil
+}
+
+func (p *pay) Pay(ctx context.Context, req payd.PayRequest) (*payd.PaymentACK, error) {
```
Comment (theflyingcodr) [inline service/pay.go:66]: Comment this and also add a few comments through the function explaining the different sections, or split it up a wee bit.

File: service/proofs.go [hunk 3]
```diff
@@ -16,9 +16,9 @@ type proofs struct {
 }
 
 // NewProofsService will setup and return a new merkle proof service.
-func NewProofsService(wtr payd.ProofsWriter) *proofs {
+func NewProofsService(rdrWtr payd.ProofsWriter) *proofs {
 	return &proofs{
-		wtr: wtr,
+		wtr: rdrWtr,
```
Comment (theflyingcodr) [inline service/proofs.go:21]: why rename to rdrwtr if it's just a Writer?

File: transaction.go [hunk 4]
```diff
@@ -83,4 +83,5 @@ type TransactionWriter interface {
 	TransactionCreate(ctx context.Context, req TransactionCreate) error
 	// TransactionUpdateState can be used to change a tx state (failed, broadcast).
 	TransactionUpdateState(ctx context.Context, args TransactionArgs, req TransactionStateUpdate) error
+	TransactionChangeCreate(ctx context.Context, txArgs TransactionCreate, dArgs DestinationCreate) error
```
Comment (theflyingcodr) [inline transaction.go:86]: Anyway we could remove this and use the existing transactioncreate and destinationscreate calls?

File: transports/http/pay.go [hunk 5]
```diff
@@ -0,0 +1,44 @@
...
+}
+
+// NewPayHandler returns a new handler for pay endpoints.
+func NewPayHandler(svc payd.PayService) *pay {
+	return &pay{
+		svc: svc,
+	}
+}
+
+func (p *pay) RegisterRoutes(g *echo.Group) {
```
Comment (theflyingcodr) [inline transports/http/pay.go:22]: Comment this

**Review by theflyingcodr — CHANGES_REQUESTED** [review]
Who needs comments 😛

File: service/proofs.go (hunk 3 above)
Comment (Tigh-Gherr) [reply service/proofs.go:21]: Ah, it was a readerwriter at one point in development, missed that when changing it back

File: transaction.go (hunk 4 above)
Comment (Tigh-Gherr) [reply transaction.go:86]: Sure I'll take a look

File: data/sqlite/transaction.go [hunk 6]
```diff
@@ -121,3 +128,52 @@ func (s *sqliteStore) TransactionUpdateState(ctx context.Context, args payd.Tran
 	return errors.Wrapf(commit(ctx, tx),
 		"failed to commit transaction when updating transactionId '%s' state to '%s'", args.TxID, req.State)
 }
+
+func (s *sqliteStore) Tx(ctx context.Context, txID string) (*bt.Tx, error) {
```
[the remark this answers is not in the document]
Comment (Tigh-Gherr) [reply data/sqlite/transaction.go:132]: This is so we conform to this interface https://github.com/libsv/go-bc/blob/master/spv/creator.go#L22

File: data/sqlite/utxo.go [hunk 7]
```diff
@@ -0,0 +1,93 @@
...
+	sqlUTXOGet = `
+	SELECT t.outpoint, t.tx_id, t.vout, d.locking_script, d.satoshis, d.derivation_path
+	FROM txos t JOIN destinations d ON t.destination_id = d.destination_id
+	WHERE reserved_for IS NULL and spent_at IS NULL and spending_txid IS NULL
+	LIMIT 0,1
+	`
+
+	sqlUTXOReserve = `
+	UPDATE txos
+	SET reserved_for = $1, updated_at = DATETIME('now')
```
[the remark this answers is not in the document]
Comment (Tigh-Gherr) [reply data/sqlite/utxo.go:22]: Good shout

File: service/pay.go (hunk 1 above)
Comment (Tigh-Gherr) [reply service/pay.go:168]: This is pointing at the payd server though.

Is the flow not: I give my callback url to the merchants P4, and then the merchant's p4 adds their own callback url to it?

Comment (theflyingcodr) [reply service/pay.go:168]: Ignore me again, this is fine yep

File: transaction.go (hunk 4 above)
Comment (Tigh-Gherr) [reply transaction.go:86]: Yep this worked

**Review by theflyingcodr — APPROVED** [review]
LGTM

**Pull request merged**